### PR TITLE
Fixes #25577 - updated puppetclass_lookup_keys_controller queries

### PR DESCRIPTION
**Pull request by ShimShtein**
1. Added `:inverse_of` to relevant objects
2. Changed queries to use less resources.

### CODE REVIEW

**Comment by mmoll** [comment]
[test katello]

File: app/controllers/puppetclass_lookup_keys_controller.rb [hunk 1]
```diff
@@ -4,8 +4,9 @@ class PuppetclassLookupKeysController < LookupKeysController
   before_action :setup_search_options, :only => :index
 
   def index
-    @lookup_keys = resource_base_search_and_page(:param_classes).smart_class_parameters
-    @puppetclass_authorizer = Authorizer.new(User.current, :collection => @lookup_keys.map {|key| key.param_class.try(:id)}.compact.uniq)
```
Comment (tbrisker) [inline app/controllers/puppetclass_lookup_keys_controller.rb:8]: Why did you remove this? afaict it is still used in the view

File: app/controllers/puppetclass_lookup_keys_controller.rb [hunk 2]
```diff
@@ -4,8 +4,9 @@ class PuppetclassLookupKeysController < LookupKeysController
...
+    @lookup_keys = resource_base_search_and_page.distinct.smart_class_parameters.preload(:lookup_values)
+    environment_classes = EnvironmentClass.where(puppetclass_lookup_key_id: @lookup_keys.map(&:id)).select(:puppetclass_id, :puppetclass_lookup_key_id).distinct.preload(:puppetclass)
+    @lookup_keys_to_class = Hash[environment_classes.map { |environment_class| [environment_class.puppetclass_lookup_key_id, environment_class.puppetclass]}]
```
Comment (tbrisker) [inline app/controllers/puppetclass_lookup_keys_controller.rb:11]: `.smart_class_parameters` already joins environment_class, so I don't think we need to toad the environment classes again - we should already have access to `environment_class.puppetclass_id` in `@lookupkeys`

**Review by tbrisker — CHANGES_REQUESTED** [review]
A couple of comments inline, can you provide some before/after numbers for this change?

File: app/controllers/puppetclass_lookup_keys_controller.rb (hunk 1 above)
Comment (ShimShtein) [reply app/controllers/puppetclass_lookup_keys_controller.rb:8]: By mistake. Now it's back

File: app/controllers/puppetclass_lookup_keys_controller.rb (hunk 2 above)
Comment (ShimShtein) [reply app/controllers/puppetclass_lookup_keys_controller.rb:11]: `.smart_class_parameters` supplies the `.joins` statement, but it does not add the environment_classes relation to the results. (We can't add it to the result, since it will mess up the pagination).
The new code preloads lookup_values with lookup keys and preloads puppetclasses with stripped down environment classes. I need the environment_class object for the lookup key connection.

**Comment by ShimShtein** [comment]
@tbrisker, updated the PR, can you review again?

File: app/controllers/puppetclass_lookup_keys_controller.rb [hunk 3]
```diff
@@ -4,8 +4,11 @@ class PuppetclassLookupKeysController < LookupKeysController
   before_action :setup_search_options, :only => :index
 
   def index
-    @lookup_keys = resource_base_search_and_page(:param_classes).smart_class_parameters
-    @puppetclass_authorizer = Authorizer.new(User.current, :collection => @lookup_keys.map {|key| key.param_class.try(:id)}.compact.uniq)
+    @lookup_keys = resource_base_search_and_page.distinct.smart_class_parameters.preload(:lookup_values)
```
Comment (tbrisker) [inline app/controllers/puppetclass_lookup_keys_controller.rb]: Since we are going directly to EnvironmentClass, why do we need `smart_class_parameters` here?

File: app/controllers/puppetclass_lookup_keys_controller.rb [hunk 4]
```diff
@@ -4,8 +4,11 @@ class PuppetclassLookupKeysController < LookupKeysController
...
+    environment_classes = EnvironmentClass.where(puppetclass_lookup_key_id: @lookup_keys.map(&:id)).select(:puppetclass_id, :puppetclass_lookup_key_id).distinct.preload(:puppetclass)
```
Comment (tbrisker) [inline app/controllers/puppetclass_lookup_keys_controller.rb:8]: Perhaps add index on environment_class(:puppetclass_lookup_key_id, :puppetclass_id)

**Comment by ShimShtein** [comment]
@tbrisker, addressed both your comments

**Comment by tbrisker** [comment]
looks like the failing test is related, but i am not sure it's a valid test anymore. @orrabin, iirc we decided to remove orphaned class params a while ago?
Edit: yes, we did, in https://github.com/theforeman/foreman/blob/develop/db/migrate/20161205142618_delete_orphaned_smart_class_parameters.rb

**Comment by ShimShtein** [comment]
@tbrisker finally the tests are passing. Can you review it again?

**Review by tbrisker — APPROVED** [review]
Thanks @ShimShtein !
